bandersnatch_vrfs/src/ring.rs

```diff
@@ -1,126 +1,128 @@
     }
-}
 
-impl KZG {
     pub fn max_keyset_size(&self) -> usize {
         self.piop_params.keyset_part_size
     }
 
     pub fn prover_key(&self, pks: Vec<SWAffine>) -> ProverKey {
-        let kzg: RealKZG = self.kzg.clone();
-        ring::piop::index::<_, CS, _>(kzg, &self.piop_params, pks).0
+        ring::index(self.pcs_params.clone(), &self.piop_params, pks).0
     }
 
     pub fn verifier_key(&self, pks: Vec<SWAffine>) -> VerifierKey {
-        let kzg: RealKZG = self.kzg.clone();
-        ring::piop::index::<_, CS, _>(kzg, &self.piop_params, pks).1
+        ring::index(self.pcs_params.clone(), &self.piop_params, pks).1
     }
-}
 
-pub type ProverKey = ring::piop::ProverKey<Fq, RealKZG, SWAffine>;
+    /// `k` is the prover secret index in [0..keyset_size).
+    pub fn init_ring_prover(&self, prover_key: ProverKey, k: usize) -> RingProver {
+        RingProver::init(prover_key, self.piop_params.clone(), k, Transcript::new(b"ring-vrf-test"))
+    }
 
-pub fn init_ring_prover(prover_key: &ProverKey, k: usize) -> RingProver {
-    ring::ring_prover::RingProver::init(prover_key, piop_params.clone(), k, Transcript::new(b"ring-vrf-test"))
+    pub fn init_ring_verifier(&self, verifier_key: VerifierKey) -> RingVerifier {
+        RingVerifier::init(verifier_key, self.piop_params.clone(), Transcript::new(b"ring-vrf-test"))
+    }
 }
 
-pub type RingProver = ring::ring_prover::RingProver<Fq, RealKZG, SWConfig>;
+impl CanonicalSerialize for KZG {
+    // Required methods
+    fn serialize_with_mode<W: Write>(
+        &self,
+        mut writer: W,
+        compress: Compress
+    ) -> Result<(), SerializationError> 
+    {
+        writer.write(&self.seed).map_err(|e| SerializationError::IoError(e))?;
+        self.pcs_params.serialize_with_mode(&mut writer, compress)
+    }
 
-pub fn ring_prove(ring_prover: &RingProver, secret: SecretBlinding<Fq,1>) -> RingProof {
-    ring_prover.prove(secret.0[0]);
+    fn serialized_size(&self, compress: Compress) -> usize {
+        32 + self.pcs_params.serialized_size(compress)
+    }
 }
 
-pub type VerifierKey = ring::piop::VerifierKey<Fq, RealKZG>;
-
-impl KZG {
-    pub fn init_ring_verifier(&self, verifier_key: &VerifierKey) -> RingVerifier {
-        ring::ring_verifier::RingVerifier::init(verifier_key, self.piop_params.clone(), Transcript::new(b"ring-vrf-test"))
+impl CanonicalDeserialize for KZG {
+    fn deserialize_with_mode<R: Read>(
+        mut reader: R,
+        compress: Compress,
+        validate: Validate
+    ) -> Result<Self, SerializationError>
+    {
+        let mut seed = [0u8; 32];
+        reader.read(&mut seed).map_err(|e| SerializationError::IoError(e))?;
+        let pcs_params = PcsParams::deserialize_with_mode(&mut reader, compress, validate)?;
+        // TODO: @jeff should we serialize the original `domain_size` to get it back here?
+        // Or shoud we use a global constant value?
+        let domain_size = 2usize.pow(10); // FIXME
```

Comment:
Here I just setted it to 2^10 (because is the value I've used within the test). 

But we allow to construct a KZG with different sizes so probably the domain size should be serialized and deserialized here

Reply:
We'll let sergey set this one, but I'll merge it anyways now.

Reply:
It's an interesting question if we should serialize the lenght.  I think here probably so, but there are several things to sort out here I guess.